presto-main/src/main/java/com/facebook/presto/metadata/SessionPropertyManager.java

```diff
@@ -61,21 +71,88 @@ public final class SessionPropertyManager
     private static final JsonCodecFactory JSON_CODEC_FACTORY = new JsonCodecFactory();
     private final ConcurrentMap<String, PropertyMetadata<?>> systemSessionProperties = new ConcurrentHashMap<>();
     private final ConcurrentMap<ConnectorId, Map<String, PropertyMetadata<?>>> connectorSessionProperties = new ConcurrentHashMap<>();
+    private final ConcurrentMap<String, PropertyMetadata<?>> workerSessionProperties = new ConcurrentHashMap<>();
+    private final Map<String, SystemSessionPropertyProvider> systemSessionPropertyProviders;
+    private final Map<String, SystemSessionPropertyProviderFactory> systemSessionPropertyProviderFactories = new ConcurrentHashMap<>();
+    private FunctionAndTypeManager functionAndTypeManager;
+    private NodeManager nodeManager;
+    private boolean nativeExecution;
+    private boolean coordinatorSidecar;
 
     public SessionPropertyManager()
     {
         this(new SystemSessionProperties());
     }
 
-    @Inject
     public SessionPropertyManager(SystemSessionProperties systemSessionProperties)
     {
         this(systemSessionProperties.getSessionProperties());
     }
 
     public SessionPropertyManager(List<PropertyMetadata<?>> systemSessionProperties)
+    {
+        this(systemSessionProperties, new ConcurrentHashMap<>());
+    }
+
+    public SessionPropertyManager(List<PropertyMetadata<?>> systemSessionProperties, Map<String, SystemSessionPropertyProvider> systemSessionPropertyProviders)
     {
         addSystemSessionProperties(systemSessionProperties);
+        this.systemSessionPropertyProviders = new ConcurrentHashMap<>(systemSessionPropertyProviders);
+    }
+
+    public SessionPropertyManager(
+            SystemSessionProperties systemSessionProperties,
+            Map<String, SystemSessionPropertyProvider> systemSessionPropertyProviders)
+    {
+        this(systemSessionProperties.getSessionProperties(), systemSessionPropertyProviders);
+    }
+
+    @Inject
+    public SessionPropertyManager(
+            SystemSessionProperties systemSessionProperties,
+            Map<String, SystemSessionPropertyProvider> systemSessionPropertyProviders,
+            FeaturesConfig featuresConfig,
+            ServerConfig serverConfig,
+            FunctionAndTypeManager functionAndTypeManager,
+            NodeManager nodeManager)
+    {
+        this(systemSessionProperties.getSessionProperties(), systemSessionPropertyProviders);
+        this.functionAndTypeManager = requireNonNull(functionAndTypeManager, "functionAndTypeManager is null");
+        this.nodeManager = requireNonNull(nodeManager, "nodeManager is null");
+        this.nativeExecution = requireNonNull(featuresConfig, "featuresConfig is null").isNativeExecutionEnabled();
+        this.coordinatorSidecar = requireNonNull(serverConfig, "serverConfig is null").isCoordinatorSidecarEnabled();
+    }
+
+    // required to load a "native" session property provider
+    public void loadSessionPropertyProvider()
```

Comment:
I think the way this should work is, we should iterate over all values in the map, and load them all.  The Presto runtime cannot make any assumptions on what plugins are loaded, except for the Java session properties.